go/teams/seitan_v2.go

```diff
@@ -0,0 +1,276 @@
+
+	sikey, err := ikey.GenerateSIKey()
+	if err != nil {
+		return pepubkey, encoded, err
+	}
+
+	keyPair, err := sikey.generateKeyPair()
+	if err != nil {
+		return pepubkey, encoded, err
+	}
+
+	var keyAndLabel keybase1.SeitanKeyAndLabelVersion2
+	keyAndLabel.K = keybase1.SeitanPubKey(keyPair.GetKID().String())
+	keyAndLabel.L = label
+
+	packedKeyAndLabel, err := libkb.MsgpackEncode(keybase1.NewSeitanKeyAndLabelWithV2(keyAndLabel))
+	if err != nil {
+		return pepubkey, encoded, err
+	}
+
+	var encKey [libkb.NaclSecretBoxKeySize]byte = secretKey
+	var naclNonce [libkb.NaclDHNonceSize]byte = nonce
+	encryptedPubKeyAndLabel := secretbox.Seal(nil, []byte(packedKeyAndLabel), &naclNonce, &encKey)
+
+	pepubkey = SeitanPEPubKey{
+		Version:              2,
+		TeamKeyGeneration:    gen,
+		RandomNonce:          nonce,
+		EncryptedKeyAndLabel: encryptedPubKeyAndLabel,
+	}
+
+	packed, err := libkb.MsgpackEncode(pepubkey)
+	if err != nil {
+		return pepubkey, encoded, err
+	}
+
+	encoded = base64.StdEncoding.EncodeToString(packed)
+	return pepubkey, encoded, nil
+}
+
+func (ikey SeitanIKeyV2) GeneratePackedEncryptedKey(ctx context.Context, team *Team, label keybase1.SeitanKeyLabel) (pepubkey SeitanPEPubKey, encoded string, err error) {
+	appKey, err := team.SeitanInviteTokenKeyLatest(ctx)
+	if err != nil {
+		return pepubkey, encoded, err
+	}
+
+	var nonce keybase1.BoxNonce
+	if _, err = rand.Read(nonce[:]); err != nil {
+		return pepubkey, encoded, err
+	}
+
+	return ikey.generatePackedEncryptedKeyWithSecretKey(appKey.Key, appKey.KeyGeneration, nonce, label)
+}
+
+func SeitanDecodePEPubKey(base64Buffer string) (pepubkey SeitanPEPubKey, err error) {
+	packed, err := base64.StdEncoding.DecodeString(base64Buffer)
+	if err != nil {
+		return pepubkey, err
+	}
+
+	err = libkb.MsgpackDecode(&pepubkey, packed)
+	return pepubkey, err
+}
+
+func (pepubkey SeitanPEPubKey) decryptKeyAndLabelWithSecretKey(secretKey keybase1.Bytes32) (ret keybase1.SeitanKeyAndLabel, err error) {
+	var encKey [libkb.NaclSecretBoxKeySize]byte = secretKey
+	var naclNonce [libkb.NaclDHNonceSize]byte = pepubkey.RandomNonce
+	plain, ok := secretbox.Open(nil, pepubkey.EncryptedKeyAndLabel, &naclNonce, &encKey)
+	if !ok {
+		return ret, errors.New("failed to decrypt seitan plain")
+	}
+
+	err = libkb.MsgpackDecode(&ret, plain)
+	if err != nil {
+		return ret, err
+	}
+
+	return ret, nil
+}
+
+func (pepubkey SeitanPEPubKey) DecryptKeyAndLabel(ctx context.Context, team *Team) (ret keybase1.SeitanKeyAndLabel, err error) {
+	appKey, err := team.SeitanInviteTokenKeyAtGeneration(ctx, pepubkey.TeamKeyGeneration)
+	if err != nil {
+		return ret, err
+	}
+
+	return pepubkey.decryptKeyAndLabelWithSecretKey(appKey.Key)
+}
+
+// "Signature"
+type SeitanSig libkb.NaclSignature
+type SeitanPubKey libkb.NaclSigningKeyPublic
```

Comment:
Maybe we should have `V2` in these type names, what do you think?

Reply:
let's leave as is -- once we deprecate v1 stuff i plan to drop the `v2` suffix too

Reply:
ah good point that we are deprecating. no bother then! thanks

Reply:
did we deprecate it yet

Reply:
I don't think so, it still works, right?